Mathlib/Algebra/GradedMonoid.lean

```diff
@@ -0,0 +1,717 @@
+
+/-- `(*) : A 0 → A 0 → A 0` is the value provided in `GradedMonoid.GMul.mul`, composed with
+an `Eq.rec` to turn `A (0 + 0)` into `A 0`.
+-/
+instance GradeZero.mul : Mul (A 0) where mul := (· • ·)
+#align graded_monoid.grade_zero.has_mul GradedMonoid.GradeZero.mul
+
+variable {A}
+
+@[simp]
+theorem mk_zero_smul {i} (a : A 0) (b : A i) : mk _ (a • b) = mk _ a * mk _ b :=
+  Sigma.ext (zero_add _).symm <| eq_rec_heq _ _ 
+#align graded_monoid.mk_zero_smul GradedMonoid.mk_zero_smul
+
+@[simp]
+theorem GradeZero.smul_eq_mul (a b : A 0) : a • b = a * b :=
+  rfl
+#align graded_monoid.grade_zero.smul_eq_mul GradedMonoid.GradeZero.smul_eq_mul
+
+end Mul
+
+section Monoid
+
+variable [AddMonoid ι] [GMonoid A]
+
+instance : Pow (A 0) ℕ where 
+  pow x n := @Eq.rec ι (n • (0:ι)) (fun a _ => A a) (GMonoid.gnpow n x) 0 (nsmul_zero n)
+
+variable {A}
+
+@[simp]
+theorem mk_zero_pow (a : A 0) (n : ℕ) : mk _ (a ^ n) = mk _ a ^ n :=
+  Sigma.ext (nsmul_zero n).symm <| eq_rec_heq _ _
+#align graded_monoid.mk_zero_pow GradedMonoid.mk_zero_pow
+
+variable (A)
+
+/-- The `Monoid` structure derived from `GMonoid A`. -/
+instance GradeZero.monoid : Monoid (A 0) :=
+  Function.Injective.monoid (mk 0) sigma_mk_injective rfl mk_zero_smul mk_zero_pow
+#align graded_monoid.grade_zero.monoid GradedMonoid.GradeZero.monoid
+
+end Monoid
+
+section Monoid
+
+variable [AddCommMonoid ι] [GCommMonoid A]
+
+/-- The `CommMonoid` structure derived from `GCommMonoid A`. -/
+instance GradeZero.commMonoid : CommMonoid (A 0) :=
+  Function.Injective.commMonoid (mk 0) sigma_mk_injective rfl mk_zero_smul mk_zero_pow
+#align graded_monoid.grade_zero.comm_monoid GradedMonoid.GradeZero.commMonoid
+
+end Monoid
+
+section MulAction
+
+variable [AddMonoid ι] [GMonoid A]
+
+/-- `GradedMonoid.mk 0` is a `MonoidHom`, using the `GradedMonoid.GradeZero.monoid` structure.
+-/
+def mkZeroMonoidHom : A 0 →* GradedMonoid A
+    where
+  toFun := mk 0
+  map_one' := rfl
+  map_mul' := mk_zero_smul
+#align graded_monoid.mk_zero_monoid_hom GradedMonoid.mkZeroMonoidHom
+
+/-- Each grade `A i` derives a `A 0`-action structure from `GMonoid A`. -/
+instance GradeZero.mulAction {i} : MulAction (A 0) (A i) :=
+  letI := MulAction.compHom (GradedMonoid A) (mkZeroMonoidHom A)
+  Function.Injective.mulAction (mk i) sigma_mk_injective mk_zero_smul
+#align graded_monoid.grade_zero.mul_action GradedMonoid.GradeZero.mulAction
+
+end MulAction
+
+end GradeZero
+
+end GradedMonoid
+
+/-! ### Dependent products of graded elements -/
+
+
+section DProd
+
+variable {α : Type _} {A : ι → Type _} [AddMonoid ι] [GradedMonoid.GMonoid A]
+
+/-- The index used by `List.dprod`. Propositionally this is equal to `(l.map fι).Sum`, but
+definitionally it needs to have a different form to avoid introducing `Eq.rec`s in `List.dprod`. -/
+def List.dProdIndex (l : List α) (fι : α → ι) : ι :=
+  l.foldr (fun i b => fι i + b) 0
+#align list.dprod_index List.dProdIndex
```

Comment:
```suggestion
def List.dprodIndex (l : List α) (fι : α → ι) : ι :=
  l.foldr (fun i b => fι i + b) 0
#align list.dprod_index List.dProdIndex
```

Reply:
@mattrobball This one is still open. Please replace all the `dProd`s by `dprod`.

Reply:
I thought dependent product shortens to DProd when a type and dProd when a term. Does the not agree with naming4?

Reply:
I have some other naming in the file based off this understanding of the new rules.

Reply:
aah, my bad. You're right!